src/libxively/event_dispatcher/xi_time_event.c

```diff
@@ -0,0 +1,348 @@
+/* Copyright (c) 2003-2015, LogMeIn, Inc. All rights reserved.
+ * This is part of Xively C library. */
+
+#include "xi_time_event.h"
+
+/**
+ * @brief This part of the file implements time event functionality. This
+ * implementation assumes that the element type is always the xi_time_event_t.
+ *
+ * It uses the vector as a container type. The vector stores pointers to the time events.
+ * Time events are always sorted in the vector by the time event
+ *
+ */
+
+/**
+ * @brief xi_swap_time_events
+ *
+ * Swaps two vector xi_time_event_t elements pointed by lhs_index and rhs_index indexes
+ * with each other. It updates element's positions value to the new ones.
+ *
+ * @param vector
+ * @param lhs_index
+ * @param rhs_index
+ */
+static void xi_swap_time_events( xi_vector_t* vector,
+                                 xi_vector_index_type_t lhs_index,
+                                 xi_vector_index_type_t rhs_index )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+    assert( lhs_index < vector->elem_no );
+    assert( rhs_index < vector->elem_no );
+
+    xi_time_event_t* lhs_time_event =
+        ( xi_time_event_t* )vector->array[lhs_index].selector_t.ptr_value;
+    xi_time_event_t* rhs_time_event =
+        ( xi_time_event_t* )vector->array[rhs_index].selector_t.ptr_value;
+
+    xi_vector_swap_elems( vector, lhs_index, rhs_index );
+
+    lhs_time_event->position = rhs_index;
+    rhs_time_event->position = lhs_index;
+}
+
+/**
+ * @brief xi_time_event_bubble_and_sort_down
+ *
+ * Helper function used for insertion and re-insertion ( cancel opertaion ). It uses the
+ * bubble sort idea to put the element on the proper position. In order to leave the
```

Comment:
idea == algorithm

I think you can remove "In order to leave the container sorted" comment, too